### ISSUE TITLE
Surface extension versions in the UI, include a simple way to update them as well?

### ISSUE THREAD

**Issue by JohnVidler** [microsoft/pxt-microbit#5143]
This has come from requests at the university for studies; these often have fast-moving custom extensions and by the time the study is live there may well be one or more that need updating for the users. Currently we do this via re-entering the extension URL into the extensions UI and re-downloading it to the project, but this could be made much cleaner with a button on the extension UI. Quick mockup of some options for this are below:

<img width="909" alt="image" src="https://github.com/microsoft/pxt-microbit/assets/537325/967ba3d2-3aaf-4605-9b9a-33cabaca5d37">

From left to right:

1. Just some text in the extended info block to say what version we're on for an up-to-date extension
2. A text link (or button?) to update the extension if the current one lags behind the latest release
3. A dropdown with all releases for this extension listed so the user can pick whichever one they want

If I had to pick the mockup I personally like the most it would be a combination of 1 and 2, depending on the state of the extension, maybe with up-to-date extensions just showing their version string (v1.2.3) on the left?

To summarise:
- Updating extensions isn't surfaced in a clean way
- Versions for extensions are only visisble in the project settings _as text_ which is quite hidden away
- Perhaps show the version on any installed extension
- Perhaps also add an easy way to jump the user to the latest one

**Comment by JohnVidler** [comment]
More information:

From talking to @abchatra in a call, the way this manifests is dependent on if you're using an official or github extension, but personally I think having some visibiltiy on pending/available updates for both would be good.

So; normally an official extension should update itself at the next opportunity (a reload, tab switch, etc.) but unofficial ones - as are the dominant case for research/experimental work - are cached and not checked as often. This means there are kiiiinda two problems (for our study use cases):

1. The visibility; the researchers/teachers conducting the study are frequently not technical experts, so following the manual update process (re-adding the URL to extensions, or updating the settings json) can be onerous.
2. The caching; the cache means that even if the extensions are manually re-added via the usual methods, they may _still_ refuse to actually update and have users be left with the old version.

Supplimentary to 2. is that weather these operations succeed or fail (either an update is performed, or the user just gets the existing code) the user experience is the same; the user is dropped back into the editor with no feedback on what actually happened, leading to more complicated processes for checking. Re-checking the project settings text file seems to be the only way to know with any certainty, but even then we've had a few cases where that also seemed to be cached and would list old versions despite the blocks listed on the main editor only existing in future versions of an extension.

I agree that we should absolutely cache extensions, but would it be possible to have some kind of rate-limited user control that allows the user to check for updates manually in a more friendly manner? Perhaps this could be implemented as a client-side check against the current-latest github tag for an extension repository, so the request load is on the client rather than the backend?

**Comment by JohnVidler** [comment]
Did a bit of a weekend hack project, and managed to get this ui, which kiiinda works - will refine when I get time.

<img width="1412" alt="image" src="https://github.com/microsoft/pxt-microbit/assets/537325/e076bd3b-2366-4512-bfaf-66be68a9c48b">

**Comment by abchatra** [comment]
@JohnVidler Github will rate limit if we do client side call for all 200+ approved extensions we have. We need to discuss the feasibility here.

**Comment by JohnVidler** [comment]
Yeah, it's not going to work if we do this for _everything_, but perhaps checking the currently loaded ones for a single project you'll be reasonable.

At the moment my hacky prototype makes 1 request per GitHub-type extension when the user opens the "active extensions" tab, and caches the result locally. Perhaps something like this, plus maybe a rate-limited "check for updates" button would work.

Keeping a medium-term cache in browser storage (so it could be shared between instances on the same host) could also reduce the call overheads.

Musing out loud.
